qutebrowser/config/configcache.py

```diff
@@ -0,0 +1,49 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2018 Jay Kamat <jaygkamat@gmail.com>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+
+"""Implementation of a basic config cache."""
+
+from qutebrowser.config import config
+
+
+class ConfigCache():
+    """A 'high-performance' cache for the config system.
+
+    Useful for areas which call out to the config system very frequently, DO
+    NOT modify the value returned, DO NOT require per-url settings, do not
+    change frequently, and do not require partially 'expanded' config paths.
+
+    If any of these requirements are broken, you will get incorrect or slow
+    behavior.
+    """
+
+    def __init__(self) -> None:
+        self.cache = {}
+        config.instance.changed.connect(self.config_changed)
+
+    def config_changed(self, attr: str) -> None:
+        if attr in self.cache:
+            self.cache[attr] = config.instance.get(attr)
+
+    def __getitem__(self, attr: str):
```

Comment:
I don't think this is a good idea because:

1. I can't actually see this function at all in profiling (not enough samples collected), so this isn't worth optimizing at all at this point :P
1. `lru_cache` is currently implemented in python and is [just a dict anyway](https://github.com/python/cpython/blob/master/Lib/functools.py#L488), which means it's just adding cpu/memory overhead for keeping track of statistics
3. `lru_cache` cannot be cleared selectively (which is the reason why this is better than sticking a `lru_cache` on `get` for config), which means on every config setting change we would blow the entire cache which will probably add much more overhead (as the lru cache needs to get re-allocated while refilling).

I'll fix all the other issues though :P

Reply:
I always assumed that python would be the one thing that gets assurance it's c code runs, but I guess not :P. I also had read somewhere the implementation was python only, but [I guess that changed in python 3.5](https://bugs.python.org/issue14373).

I did a [small profile](http://paste.debian.net/1040719/) and for me at least there is a noticable improvement, so if we do end up seeing this in the profiles later, we definetly should revisit this (since the cost of allocation is probably a lot cheaper in C):

```
[jgkamat@vall /tmp]$ time py test.py                                                      [2166]  8:29AM
python3 test.py  1.20s user 0.01s system 99% cpu 1.212 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2167]  8:29AM
python3 test.py  1.27s user 0.01s system 99% cpu 1.274 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2167]  8:29AM
^[[Apython3 test.py  1.21s user 0.00s system 99% cpu 1.217 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2167]  8:29AM
python3 test.py  1.28s user 0.01s system 99% cpu 1.286 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2167]  8:29AM
^[[Apython3 test.py  1.21s user 0.00s system 99% cpu 1.216 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2167]  8:29AM
^[[Apython3 test.py  1.25s user 0.00s system 99% cpu 1.262 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2167]  8:29AM
python3 test.py  1.21s user 0.00s system 99% cpu 1.217 total
[jgkamat@vall /tmp]$ # lru cache                                                          [2167]  8:29AM
[jgkamat@vall /tmp]$ time py test.py                                                      [2168]  8:29AM
python3 test.py  0.99s user 0.01s system 99% cpu 0.993 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2169]  8:29AM
^[[Apython3 test.py  1.01s user 0.01s system 99% cpu 1.018 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2169]  8:29AM
python3 test.py  0.99s user 0.00s system 99% cpu 0.991 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2169]  8:29AM
python3 test.py  1.01s user 0.01s system 99% cpu 1.017 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2169]  8:29AM
python3 test.py  1.00s user 0.01s system 99% cpu 1.007 total
[jgkamat@vall /tmp]$ time py test.py                                                      [2169]  8:29AM
python3 test.py  1.01s user 0.00s system 99% cpu 1.018 total
```